lib/spack/spack/test/installer.py

```diff
@@ -510,6 +510,65 @@ def test_install_failed(install_mockery, monkeypatch, capsys):
     assert 'Warning: b failed to install' in out
 
 
+def test_install_fail_on_interrupt(install_mockery, monkeypatch, capsys):
+    """Test ctrl-c interrupted install."""
+    err_msg = 'mock keyboard interrupt'
+
+    def _interrupt(installer, task, **kwargs):
+        raise KeyboardInterrupt(err_msg)
+
+    spec, installer = create_installer('a')
+
+    # Raise a KeyboardInterrupt error to trigger early termination
+    monkeypatch.setattr(inst.PackageInstaller, '_install_task', _interrupt)
+
+    with pytest.raises(KeyboardInterrupt, match=err_msg):
+        installer.install()
+
+    out = str(capsys.readouterr())
+    assert 'Failed to install' in out
+    assert err_msg in out
+
+
+def test_install_fail_fast_on_detect(install_mockery, monkeypatch, capsys):
+    """Test fail_fast install when detect an install failure."""
+    spec, installer = create_installer('a')
+
+    # Make sure the package is identified as failed
+    monkeypatch.setattr(spack.database.Database, 'prefix_failed', _true)
+
+    # Skip the actual installation though it should never get there
+    monkeypatch.setattr(inst.PackageInstaller, '_install_task', _noop)
+
+    msg = 'Fail-fast termination of install'
+    with pytest.raises(spack.installer.InstallError, match=msg):
+        installer.install(fail_fast=True)
+
+    out = str(capsys.readouterr())
+    assert 'b failed to install' in out
+
+
+def test_install_fail_fast_on_except(install_mockery, monkeypatch, capsys):
+    """Test fail_fast install when encounter install failure."""
+    err_msg = 'mock install failure'
+
+    def _install(installer, task, **kwargs):
+        raise RuntimeError(err_msg)
+
+    spec, installer = create_installer('a')
+
+    # Raise a non-KeyboardInterrupt exception to trigger fast failure
+    monkeypatch.setattr(inst.PackageInstaller, '_install_task', _install)
```

Comment:
I was narrowly focused on coverage for the `installer.py` module.  I'll have to step back and reconsider the test from the `install` level.  Code coverage is especially challenging when the functions are (still) long.

I'm not sure why mocking other methods in a class is less than ideal when the goal is to increase coverage (in challenging functions).  What am I missing?

Reply:
> I'm not sure why mocking other methods in a class is less than ideal when the goal is to increase coverage (in challenging functions). What am I missing?

I think increasing coverage is better. But I think a different approach should be taken. I suggested overriding `Package.install` to fail.

If you are testing a class, I think it increases complexity if you mock methods in that same class. I would draw an analogy to methods that include a `testing` parameter: if the behavior of the object is changing because of the test, how do we ensure that the object is properly tested such that we are sure of its behavior outside of the testing environment. This is generally a gray area, but I think that changing the behavior of the object itself as little as possible helps reason about the test.

Does that seem reasonable?

Reply:
Why override `Package.install`?  Are you proposing a refactor?  

The check is in the `try-except` block within `PackageInstaller.install`'s `while` (have build tasks) loop.  The only things that could meaningfully be monkeypatched are the installer's `_install_task` or `_update_installed`; hence, the test monkeypatches `_install_task`.

Note that `_install_task` is roughly equivalent to a single pass -- as in processing a single spec -- of the old `PackageBase.do_install`.

I can change the test to run `Package.do_install` while still monkeypatching `_install_task` if you prefer.  However, as a unit test to cover a special case, I would think the goal is to keep the amount of code touched by the test as small as possible.

Reply:
> The only things that could meaningfully be monkeypatch'd are the installer's _install_task -- or something within it (e.g., the install phase)

The install phase was what I was specifically thinking of: you can achieve that by using the mock package `failing-build` (which just throws an `InstallError` when you try to install it). You can also possibly use an entirely-mocked replacement for `Package` with an `install` method that fails.

> I chose to monkeypatch _install_task to keep the test fairly straight-forward.

I agree that it is straightforward, but I worry that it is fragile. I think it may be fragile for the reasons I explained in https://github.com/spack/spack/pull/15295/files#r391203004. 

> I also wanted to keep the amount of code touched by the case as small as possible.

I agree that is good, but I think other criteria are more important: clarity and fragility. Right now the code is clear but I'm worried about it being fragile.

To make sure: do you not see a benefit to my suggestion, or do you think the benefits of your approach outweigh the benefits of my suggestion?

Reply:
I don't see the fragility issue.  IMO the behavior change for the test is needed because of the structure of the code.  Also, I consider this a low-level unit test in the sense of testing a very narrow condition not easily met otherwise.

Bottom line: I am saying I think the benefits of my approach outweigh those of your suggestion.

Reply:
I agree that creating a test that requires knowledge of Package phases would likely be too complex. I don't think it's too complex though to make use of a package testing resource that fails. While understanding the available phases and how Spack arranges for them to execute is also complicated, I think that there will always be a `Package.install` method, and I don't think it is too much to expect that `installer.py` will always be invoking this at some point.

> I would rather remove the test (and minimally lower coverage) than make the requested change

There is some overlap between checks on `Database.prefix_failed` and logic which handles exceptions with `Package.install`, so I agree that removing this test shouldn't reduce coverage (certainly from a logic perspective if not a line-coverage perspective). I like the idea of testing both situations though.

My own summary: testing what happens when a package install fails should be simple (not just for this test but for all Spack tests). If it's not, then I should make it simple.

Reply:
Are you saying a basic unit test intended to ensure errors are not in the exception clause of a `try-catch` block (e.g., formats don't raise exceptions) is not as valid to retain as an integration test that you can reason about?

Reply:
No: I think they are both valid tests. I meant that I thought `test_install_fail_fast_on_except` and `test_install_fail_fast_on_detect` have some logical overlap and so it isn't as critical that both exist. 

As I think about it more, I think I'm wrong about that:

* `test_install_fail_fast_on_except` checks behavior when an attempt to install the package fails
* `test_install_fail_fast_on_detect` checks behavior when it is discovered that some other process failed to install the package

So in short: both tests are good and should be kept. I still think this test should change (possibly after the testing framework is refactored to make it easier to test failing package installs).

Reply:
@scheibelp this does not seem too fragile to me. It clearly monkeypatches a single function to fail, and that function is integral to the API, so it cannot change to break the test without complete revision of the installer, at which point rewriting tests is appropriate.

Reply:
> The install phase was what I was specifically thinking of: you can achieve that by using the mock package `failing-build` (which just throws an `InstallError` when you try to install it). You can also possibly use an entirely-mocked replacement for `Package` with an `install` method that fails.

Using any package's overidden `install` method is not going to exercise the path that is being tested with `test_install_fail_fast_on_except`.  That path can only occur with an exception that is raised from either `_install_task` or `_update_installed`.  I will have to look at other modules I could patch.

Reply:
It looks like the code has since been updated to resolve this.